pkg/operator/ceph/pool/radosnamespace/controller.go

```diff
@@ -0,0 +1,316 @@
+		r.updateStatus(r.client, request.NamespacedName, cephv1.ConditionProgressing)
+	}
+
+	// Make sure a CephCluster is present otherwise do nothing
+	cephCluster, isReadyToReconcile, cephClusterExists, reconcileResponse := opcontroller.IsReadyToReconcile(r.opManagerContext, r.client, request.NamespacedName, controllerName)
+	if !isReadyToReconcile {
+		// This handles the case where the Ceph Cluster is gone and we want to delete that CR
+		// We skip the deleteRadosNamespace() function since everything is gone already
+		//
+		// Also, only remove the finalizer if the CephCluster is gone
+		// If not, we should wait for it to be ready
+		// This handles the case where the operator is not ready to accept Ceph command but the cluster exists
+		if !cephBlockPoolRadosNamespace.GetDeletionTimestamp().IsZero() && !cephClusterExists {
+			// Remove finalizer
+			err = opcontroller.RemoveFinalizer(r.opManagerContext, r.client, cephBlockPoolRadosNamespace)
+			if err != nil {
+				return opcontroller.ImmediateRetryResult, errors.Wrap(err, "failed to remove finalizer")
+			}
+
+			// Return and do not requeue. Successful deletion.
+			return reconcile.Result{}, nil
+		}
+		return reconcileResponse, nil
+	}
+
+	// Populate clusterInfo during each reconcile
+	r.clusterInfo, _, _, err = mon.LoadClusterInfo(r.context, r.opManagerContext, request.NamespacedName.Namespace)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to populate cluster info")
+	}
+	r.clusterInfo.Context = r.opManagerContext
+
+	// DELETE: the CR was deleted
+	if !cephBlockPoolRadosNamespace.GetDeletionTimestamp().IsZero() {
+		logger.Debugf("delete cephBlockPoolRadosNamespace %q", namespacedName)
+		// On external cluster, we don't delete the rados namespace, it has to be deleted manually
+		if cephCluster.Spec.External.Enable {
+			logger.Warning("external rados namespace %q deletion is not supported, delete it manually", namespacedName)
+		} else {
+			err := r.deleteRadosNamespace(cephBlockPoolRadosNamespace)
+			if err != nil {
+				if strings.Contains(err.Error(), opcontroller.UninitializedCephConfigError) {
+					logger.Info(opcontroller.OperatorNotInitializedMessage)
+					return opcontroller.WaitForRequeueIfOperatorNotInitialized, nil
+				}
+				return reconcile.Result{}, errors.Wrapf(err, "failed to delete ceph blockpool rados namespace %q", cephBlockPoolRadosNamespace.Name)
+			}
+		}
+		err = csi.SaveClusterConfig(r.context.Clientset, buildClusterID(cephBlockPoolRadosNamespace), r.clusterInfo, nil)
+		if err != nil {
+			return reconcile.Result{}, errors.Wrap(err, "failed to save cluster config")
+		}
+		// Remove finalizer
+		err = opcontroller.RemoveFinalizer(r.opManagerContext, r.client, cephBlockPoolRadosNamespace)
+		if err != nil {
+			return reconcile.Result{}, errors.Wrap(err, "failed to remove finalizer")
+		}
+
+		// Return and do not requeue. Successful deletion.
+		return reconcile.Result{}, nil
+	}
+
+	if !cephCluster.Spec.External.Enable {
```

Comment:
The rados namespace behaves so differently for external clusters, how about refactoring this into two methods to handle external or local? The multiple checks for external clusters is confusing to me. Or just error for external clusters as suggested below?

Reply:
Followed the same code pattern used for https://github.com/rook/rook/blob/master/pkg/operator/ceph/file/subvolumegroup/controller.go#L211-L271 if required to separate it out i can do it. please let me know what do you think

Reply:
I think the refactor would really help understand the code flow, something like:
```
if external cluster {
  // call a helper method to update the configmap
  return
}
// for local clusters create the namespace 
// call helper method to update the configmap

Reply:
Done. added help function and did refactoring.